spec/services/container_deployment_service_spec.rb

```diff
@@ -0,0 +1,41 @@
+require 'recursive-open-struct'
+
+describe ContainerDeploymentService do
+  before do
+    %w(amazon openstack google azure redhat vmware).each do |p|
+      network = FactoryGirl.create(:network, :ipaddress => "127.0.0.1")
+      hardware = FactoryGirl.create(:hardware)
+      hardware.networks << network
+      vm = FactoryGirl.create("vm_#{p}".to_sym, :hardware => hardware)
+      if (p.include? "amazon") || (p.include? "redhat")
+        template = FactoryGirl.create("template_#{p}".to_sym)
+        provider = FactoryGirl.create("ems_#{p}".to_sym, :miq_templates => [template])
+      else
+        provider = FactoryGirl.create("ems_#{p}".to_sym)
+      end
+      provider.vms << vm
+    end
+    @foreman_provider = FactoryGirl.create(:configuration_manager_foreman)
+  end
+
+  context "possible_providers_and_vms" do
+    it "finds all Cloud and Infra providers and their existing VMs" do
+      providers = ContainerDeploymentService.new.possible_providers_and_vms
+      vms = providers.collect_concat { |p| p[:vms] }
+      expect(providers.size).to eq(6)
+      expect(vms.size).to eq(6)
```

Comment:
> vms is built from providers can it be of different size ? what do we check by checking it size ?

If there's something wrong with the service it might not return all the cloud/infra providers and their vms. in this case I created 6 vms, one per provider and checking that the service returned all 6.